File: backend/app/api/automl.py
"""自动化机器学习 API — CSV 上传、模型目录、训练 run。"""

from __future__ import annotations

from typing import Annotated
from urllib.parse import quote

from fastapi import APIRouter, Depends, File, Query, UploadFile
from fastapi.responses import FileResponse
from sqlalchemy.orm import Session

from app.api.deps import get_client_ip, get_current_user, require_feature
from app.database import get_db
from app.models.org import User
from app.schemas.automl import (
    AutoMlDatasetUploadOut,
    AutoMlMetaOut,
    AutoMlModelsOut,
    AutoMlRunCreateIn,
    AutoMlRunOut,
    AutoMlRunSummaryOut,
)
from app.schemas.common import ApiResponse
from app.services import automl_service as svc
from app.services.audit_service import write_audit

router = APIRouter(
    prefix="/auto-ml",
    tags=["auto-ml"],
    dependencies=[Depends(require_feature("auto_ml"))],
)


@router.get("/meta", response_model=ApiResponse[AutoMlMetaOut])
def auto_ml_meta(
    _: Annotated[User, Depends(get_current_user)],
) -> ApiResponse[AutoMlMetaOut]:
    return ApiResponse(data=svc.get_meta())


@router.post("/datasets/upload", response_model=ApiResponse[AutoMlDatasetUploadOut])
async def upload_dataset(
    user: Annotated[User, Depends(get_current_user)],
    db: Annotated[Session, Depends(get_db)],
    file: UploadFile = File(..., description="CSV data file"),
    client_ip: Annotated[str | None, Depends(get_client_ip)] = None,
) -> ApiResponse[AutoMlDatasetUploadOut]:
    content = await file.read()
    result = svc.upload_dataset(
        user_id=user.id,
        filename=file.filename or "data.csv",
        content=content,
    )
    write_audit(
        db,
        user_id=user.id,
        action="auto_ml.upload",
        resource_type="auto_ml",
        detail={
            "dataset_id": result.dataset_id,
            "filename": file.filename,
            "bytes": len(content),
        },
        ip_address=client_ip,
    )
    return ApiResponse(data=result)


@router.get("/datasets/{dataset_id}", response_model=ApiResponse[AutoMlDatasetUploadOut])
def get_dataset(
    dataset_id: str,
    user: Annotated[User, Depends(get_current_user)],
) -> ApiResponse[AutoMlDatasetUploadOut]:
    return ApiResponse(data=svc.get_dataset(user_id=user.id, dataset_id=dataset_id))


@router.get("/models", response_model=ApiResponse[AutoMlModelsOut])
def list_models(
    user: Annotated[User, Depends(get_current_user)],
    task: str = Query(..., description="classification|regression|clustering|anomaly|time_series"),
) -> ApiResponse[AutoMlModelsOut]:
    _ = user
    return ApiResponse(data=svc.list_models(task))


@router.post("/runs", response_model=ApiResponse[AutoMlRunOut])
def create_run(
    body: AutoMlRunCreateIn,
    user: Annotated[User, Depends(get_current_user)],
    db: Annotated[Session, Depends(get_db)],
    client_ip: Annotated[str | None, Depends(get_client_ip)] = None,
) -> ApiResponse[AutoMlRunOut]:
    result = svc.create_run(db, user_id=user.id, body=body)
    write_audit(
        db,
        user_id=user.id,
        action="auto_ml.run.create",
        resource_type="auto_ml",
        detail={"run_id": result.run_id, "task": body.task, "dataset_id": body.dataset_id},
        ip_address=client_ip,
    )
    return ApiResponse(data=result)


@router.get("/runs", response_model=ApiResponse[list[AutoMlRunSummaryOut]])
def list_runs(
    user: Annotated[User, Depends(get_current_user)],
    limit: int = Query(30, ge=1, le=100),
) -> ApiResponse[list[AutoMlRunSummaryOut]]:
    return ApiResponse(data=svc.list_runs(user_id=user.id, limit=limit))


@router.get("/runs/{run_id}", response_model=ApiResponse[AutoMlRunOut])
def get_run(
    run_id: str,
    user: Annotated[User, Depends(get_current_user)],
) -> ApiResponse[AutoMlRunOut]:
    return ApiResponse(data=svc.get_run(user_id=user.id, run_id=run_id))


@router.get("/runs/{run_id}/download/predictions")
def download_predictions(
    run_id: str,
    user: Annotated[User, Depends(get_current_user)],
) -> FileResponse:
    path = svc.predictions_path(user_id=user.id, run_id=run_id)
    filename = f"automl_{run_id[:8]}_predictions.csv"
    return FileResponse(
        path,
        media_type="text/csv",
        filename=filename,
        headers={
            "Content-Disposition": f"attachment; filename*=UTF-8''{quote(filename)}"
        },
    )


@router.get("/runs/{run_id}/download/model")
def download_model(
    run_id: str,
    user: Annotated[User, Depends(get_current_user)],
) -> FileResponse:
    path = svc.model_path(user_id=user.id, run_id=run_id)
    filename = f"automl_{run_id[:8]}_{path.name}"
    return FileResponse(
        path,
        media_type="application/octet-stream",
        filename=filename,
        headers={
            "Content-Disposition": f"attachment; filename*=UTF-8''{quote(filename)}"
        },
    )

File: frontend/src/api/automl.js
/** 自动化机器学习 REST API */
import { api, getApiBase, getToken, rejectHttpFailure } from "./http.js";
import { downloadBlob } from "../utils/downloadBlob.js";

export async function fetchAutoMlMeta() {
  return api("/api/v1/auto-ml/meta");
}

export async function uploadAutoMlDataset(file) {
  const form = new FormData();
  form.append("file", file);
  return api("/api/v1/auto-ml/datasets/upload", { method: "POST", body: form });
}

export async function fetchAutoMlDataset(datasetId) {
  return api(`/api/v1/auto-ml/datasets/${datasetId}`);
}

export async function fetchAutoMlModels(task) {
  return api(`/api/v1/auto-ml/models?task=${encodeURIComponent(task)}`);
}

export async function createAutoMlRun(body) {
  return api("/api/v1/auto-ml/runs", { method: "POST", body });
}

export async function fetchAutoMlRun(runId) {
  return api(`/api/v1/auto-ml/runs/${runId}`);
}

export async function listAutoMlRuns(limit = 30) {
  return api(`/api/v1/auto-ml/runs?limit=${limit}`);
}

async function downloadAutoMlFile(url, fallbackName) {
  const res = await fetch(`${getApiBase()}${url}`, {
    headers: { Authorization: `Bearer ${getToken()}` },
  });
  if (!res.ok) {
    const json = await res.json().catch(() => ({}));
    rejectHttpFailure(res, json);
  }
  const blob = await res.blob();
  const disp = res.headers.get("Content-Disposition") || "";
  const m = /filename\*=UTF-8''([^;]+)|filename="?([^";]+)"?/i.exec(disp);
  const name = decodeURIComponent(m?.[1] || m?.[2] || fallbackName);
  downloadBlob(blob, name);
}

export async function downloadAutoMlPredictions(runId) {
  await downloadAutoMlFile(
    `/api/v1/auto-ml/runs/${runId}/download/predictions`,
    `automl_${String(runId).slice(0, 8)}_predictions.csv`
  );
}

export async function downloadAutoMlModel(runId) {
  await downloadAutoMlFile(
    `/api/v1/auto-ml/runs/${runId}/download/model`,
    `automl_${String(runId).slice(0, 8)}_model.pkl`
  );
}
